tools/cache_logs_player.py

```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python3
+import argparse
+from datetime import datetime
+import aioredis
+import asyncio
+from aiocsv import AsyncReader
+import aiofiles
+
+'''
+To install: pip install -r requirements.txt
+'''
+
+
+class Command:
+    args = None
+    client_id = 0
+
+class TwitterCacheTraceParser:
+    """
+    https://github.com/twitter/cache-trace
+    """
+    def parse(self, csv) -> Command:
+        operation = csv[5]
+        key = csv[1] + "a"
+        value_size = int(csv[3])
+        synthetic_value = "".zfill(value_size)
+
+        client_id = csv[4]
+        ttl = csv[6]
+
+        cmd = Command()
+        cmd.client_id = client_id
+
+        if operation == "get":
+            cmd.args = ["GET", key]
+        elif operation == 'gets':
+            cmd.args = ["GET", key]
+        elif operation == 'set':
+            cmd.args = ["SET", key, synthetic_value]
+        elif operation == 'add':
+            cmd.args = ["SET", key, synthetic_value]
+        elif operation == 'replace':
+            cmd.args = ["SET", key, synthetic_value]
+        elif operation == 'cas':
+            cmd.args = ["SET", key, synthetic_value]
+        elif operation == 'append':
+            cmd.args = ["SET", key, synthetic_value]
+        elif operation == 'prepend':
+            cmd.args = ["SET", key, synthetic_value]
+        elif operation == 'delete':
+            cmd.args = ["DEL", key]
+        elif operation == 'incr':
+            cmd.args = ["INCR", key]
+        elif operation == 'decr':
+            cmd.args = ["DECR", key]
+
+        return cmd
+
+class AsyncWorker:
+    QUEUE_SIZE = 100000
+    BATCH_SIZE = 20
+
+    def __init__(self, redis_client) -> None:
+        self.queue = asyncio.Queue(self.QUEUE_SIZE)
+        self.redis_client = redis_client
+        self.working = False
+
+    async def put(self, cmd: Command) -> None:
+        await self.queue.put(cmd)
+
+    async def work(self) -> None:
+        self.working = True
+        batch = []
+        while self.working or len(batch) > 0 or not self.queue.empty() :
+            try:
+                cmd = await asyncio.wait_for(self.queue.get(), timeout=1.0)
+                batch.append(cmd)
+                if len(batch) >= self.BATCH_SIZE:
+                    await self.execute(batch)
+                    batch.clear()
```

Comment:
If you group the values on the producer side instead and send them over the channel in batches you would reduce the pressure on the asyncio scheduler

But I suspect its not a bottleneck here, besides managing the timeout to send half filled batches from the producer would be a little inconvenient

Reply:
@dranikpg thanks, good feedback.